apps/site/pages/en/learn/typescript/publishing-a-ts-package.md

```diff
@@ -0,0 +1,332 @@
+
+- `.nvmrc` comes from [NVM](https://github.com/nvm-sh/nvm), a multi-version manager for node. It allows you to specify the version of node the project should generally use.
+
+A directory overview of a repository would look something like:
+
+```text displayName="Files co-located"
+example-ts-pkg/
+├ .github/
+│ ├ workflows/
+│ │ ├ ci.yml
+│ │ └ publish.yml
+│ └ dependabot.yml
+├ src/
+│ ├ foo.fixture.js
+│ ├ main.ts
+│ ├ main.test.ts
+│ ├ some-util.ts
+│ └ some-util.test.ts
+├ LICENSE
+├ package.json
+├ README.md
+└ tsconfig.json
+```
+
+```text displayName="Files co-located but segregated"
+example-ts-pkg/
+├ .github/
+│ ├ workflows/
+│ │ ├ ci.yml
+│ │ └ publish.yml
+│ └ dependabot.yml
+├ src/
+│ ├ __test__/
+│ │ ├ foo.fixture.js
+│ │ ├ main.test.ts
+│ ├ main.ts
+│ └ some-util.ts
+│ │ ├ __test__
+│ │   └ some-util.test.ts
+│ │ └ some-util.ts
+├ LICENSE
+├ package.json
+├ README.md
+└ tsconfig.json
+```
+
+```text displayName="'src' and 'test' fully segregated"
+example-ts-pkg/
+├ .github/
+│ ├ workflows/
+│ │ ├ ci.yml
+│ │ └ publish.yml
+│ └ dependabot.yml
+├ src/
+│ ├ main.ts
+│ ├ some-util.ts
+├ test/
+│ ├ foo.fixture.js
+│ ├ main.ts
+│ └ some-util.ts
+├ LICENSE
+├ package.json
+├ README.md
+└ tsconfig.json
+```
+
+And a directory overview of its published package would look something like:
+
+```text displayName="Fully flat"
+example-ts-pkg/
+├ LICENSE
+├ main.d.ts
+├ main.d.ts.map
+├ main.js
+├ package.json
+├ README.md
+├ some-util.d.ts
+├ some-util.d.ts.map
+└ some-util.js
+```
+
+```text displayName="With 'dist'"
+example-ts-pkg/
+├ dist/
+│ ├ main.d.ts
+│ ├ main.d.ts.map
+│ ├ main.js
+│ ├ some-util.d.ts
+│ ├ some-util.d.ts.map
+│ └ some-util.js
+├ LICENSE
+├ package.json
+└ README.md
+```
+
+A note about directory organisation: There are a few common practices for placing tests. Principle of least knowledge says to co-locate them (put them adjacent to implementation). Sometimes, that's in the same directory, or within a drawer like a `__test__` (also adjacent to the implementation, "Files co-located but segregated"). Alternatively, some opt to create a `test/` sibling to `src/` ("'src' and 'test' fully segregated"), either with a mirrored structure or a "junk drawer".
+
+## What to do with your types
+
+### Treat types like a test
+
+The purpose of types is to warn an implementation will not work:
+
+```ts
+const foo = 'a';
+const bar: number = 1 + foo;
+//    ^^^ Type 'string' is not assignable to type 'number'.
+```
+
+TypeScript has warned that the above code will not behave as intended, just like a unit test warns that code does not behave as intended. They are complementary and verify different things—you should have both.
+
+Your editor (ex VS Code) likely has built-in support for TypeScript, displaying errors as you work. If not, and/or you missed those, CI will have your back.
```

Comment:
```suggestion
Your editor (eg VS Code) likely has built-in support for TypeScript, displaying errors as you work. If not, and/or you missed those, CI will have your back.
```

Reply:
These are the same 😜 One is english, the other is latin:

`ex` → `example`
`eg` → [`exempli gratia`](https://en.wikipedia.org/wiki/Exempli_gratia)

Reply:
right, but "ex" reads to me as "formerly", "eg" doesn't have the same confusion

Reply:
Ah, okay that's better then 🙂